src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs

```diff
@@ -0,0 +1,208 @@
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+
+namespace Infrastructure.Common
+{
+    // This class manages the adding and removal of certificates.
+    // It also handles informing http.sys of which certificate to associate with SSL ports.
+    internal static class CertificateManager
+    {
+        private static object s_certificateLock = new object();
+        private static bool s_PlatformSpecificStoreLocationIsSet = false;
+        private static StoreLocation s_PlatformSpecificRootStoreLocation = StoreLocation.LocalMachine;
+
+        // The location from where to open StoreName.Root
+        //
+        // Since we don't have access to System.Runtime.InteropServices.IsOSPlatform API, we need to find a novel way to switch
+        // the store we use on Linux
+        //
+        // For Linux: 
+        // On Linux, opening stores from the LocalMachine store is not supported yet, so we toggle to use CurrentUser
+        // Furthermore, we don't need to sudo in Linux to install a StoreName.Root : StoreLocation.CurrentUser. So this will allow
+        // tests requiring certs to pass in Linux. 
+        // See dotnet/corefx#3690
+        //
+        // For Windows: 
+        // We don't want to use CurrentUser, as writing to StoreName.Root : StoreLocation.CurrentUser will result in a 
+        // modal dialog box that isn't dismissable if the root cert hasn't already been installed previously. 
+        // If the cert has already been installed, writing to StoreName.Root : StoreLocation.CurrentUser results in a no-op
+        // 
+        // In other words, on Windows, we can bypass the modal dialog box, but only if we install to StoreName.Root : StoreLocation.LocalMachine
+        // To do this though means that we must run certificate-based tests elevated
+        private static StoreLocation PlatformSpecificRootStoreLocation
+        {
+            get
+            {
+                if (!s_PlatformSpecificStoreLocationIsSet)
+                {
+                    try
+                    {
+                        using (var store = new X509Store(StoreName.Root, StoreLocation.LocalMachine))
+                        {
+                            store.Open(OpenFlags.ReadWrite);
+                        }
+                    }
+                    catch (PlatformNotSupportedException)
+                    {
+                        // Linux
+                        s_PlatformSpecificRootStoreLocation = StoreLocation.CurrentUser; 
+                    }
+                    s_PlatformSpecificStoreLocationIsSet = true;
+                }
+                return s_PlatformSpecificRootStoreLocation;
+            }
+        }
+
+        // Adds the given certificate to the given store unless it is
+        // already present.  Returns the  certificate either already in
+        // the store or the one requested.
+        public static X509Certificate2 AddToStoreIfNeeded(StoreName storeName,
```

Comment:
Reviewers: almost verbatim from BridgeClientCertificateManager.  I removed the separate cache we were maintaining and made it return the cert actually in the store if one was already present.